### DISCUSSION TITLE
debugging Mitsuba xml scene import into Blender

### DISCUSSION THREAD

**Discussion by tadz-io** [mitsuba-renderer/mitsuba-blender#101] [Q&A] [answered]
I'm following Mitsuba's caustic rendering tutorial: https://mitsuba.readthedocs.io/en/latest/src/inverse_rendering/caustics_optimization.html . Now i wanted to export the scene described in the tutorial as a xml file and import it into Blender to visualize the setup. The scene is a python dict (see tutorial for further details):

``` python
scene = {
    'type': 'scene',
    'sensor': sensor,
    'integrator': integrator,
    # Glass BSDF
    'simple-glass': {
        'type': 'dielectric',
        'id': 'simple-glass-bsdf',
        'ext_ior': 'air',
        'int_ior': 1.5,
        'specular_reflectance': { 'type': 'spectrum', 'value': 0 },
    },
    'white-bsdf': {
        'type': 'diffuse',
        'id': 'white-bsdf',
        'reflectance': { 'type': 'rgb', 'value': (1, 1, 1) },
    },
    'black-bsdf': {
        'type': 'diffuse',
        'id': 'black-bsdf',
        'reflectance': { 'type': 'spectrum', 'value': 0 },
    },
    # Receiving plane
    'receiving-plane': {
        'type': 'obj',
        'id': 'receiving-plane',
        'filename': 'meshes/rectangle.obj',
        'to_world': \
            mi.ScalarTransform4f.look_at(
                target=[0, 1, 0],
                origin=[0, -7, 0],
                up=[0, 0, 1]
            ).scale((5, 5, 5)),
        'bsdf': {'type': 'ref', 'id': 'white-bsdf'},
    },
    # Glass slab, excluding the 'exit' face (added separately below)
    'slab': {
        'type': 'obj',
        'id': 'slab',
        'filename': 'meshes/slab.obj',
        'to_world': mi.ScalarTransform4f.rotate(axis=(1, 0, 0), angle=90),
        'bsdf': {'type': 'ref', 'id': 'simple-glass'},
    },
    # Glass rectangle, to be optimized
    'lens': {
        'type': 'ply',
        'id': 'lens',
        'filename': lens_fname,
        'to_world': mi.ScalarTransform4f.rotate(axis=(1, 0, 0), angle=90),
        'bsdf': {'type': 'ref', 'id': 'simple-glass'},
    },

    # Directional area emitter placed behind the glass slab
    'focused-emitter-shape': {
        'type': 'obj',
        'filename': 'meshes/rectangle.obj',
        'to_world': mi.ScalarTransform4f.look_at(
            target=[0, 0, 0],
            origin=[0, 5, 0],
            up=[0, 0, 1]
        ),
        'bsdf': {'type': 'ref', 'id': 'black-bsdf'},
        'focused-emitter': emitter,
    },
}
```

i have exported the dict to a xml using Mitsuba:
``` python
import mitsuba as mi
mi.xml.dict_to_xml(scene, filename='caustic_optimization.xml')
```

I'm running into several issues, some of which i could solve by manually editing the xml file. However, for the more complex issues i would like to debug the python code of `mitsuba-blender` . I'm using VSC, i have cloned this repo and created a virtual environment using python 3.11. However installing `bpy` gives me the following error:

```
pip install bpy       
ERROR: Ignored the following versions that require a different python version: 2.82.1 Requires-Python >=3.7, <3.8
ERROR: Could not find a version that satisfies the requirement bpy (from versions: none)
ERROR: No matching distribution found for bpy
```

Downgrading python to 3.7.x won't solve it because of other dependency issues... Long story short: how can i setup an environment to run some tests on importing mitsuba xml's into Blender?

I'm using Blender 3.6

**Comment by bathal1** [comment, marked as answer]
Hi,

Installing `bpy` via pip will not help you. You need to use blender's python, which already has `bpy` installed (since it's blender's python API).

For development, you can clone the repo and make a symlink to the `mitsuba-blender` folder in the directory containing add-ons for blender, e.g. for linux: ` ~/.config/blender/3.6/scripts/addons`. Alternatively you could use [this ](https://marketplace.visualstudio.com/items?itemName=JacquesLucke.blender-development) VSCode extension.

**Reply by tadz-io** [reply]
Thanks for the quick response. Much appreciated!

Yes i tried using Blender's python interpreter as well. I installed pytest and tried running the tests. But all tests fail and many of them complain that `bpy.ops.import_scene.mitsuba` can't be found. Example below: 
```
./tests/test_compare.py::test_round_trip[scenes/test1.xml] Failed: [undefined]AttributeError: Calling operator "bpy.ops.import_scene.mitsuba" error, could not be found
xml_scene = 'scenes/test1.xml'
resource_resolver = <fixtures.ResourceResolver object at 0x12611fbb0>
mitsuba_scene_ztest = <fixtures.MitsubaRenderTester object at 0x12611faf0>
```

Blender's python interpreter is using bpy 4.0.0. I tried upgrading to 4.1.0 but no luck:

```
/Applications/Blender.app/Contents/Resources/3.6/python/bin/python3.10 -m pip install bpy==4.1.0  
Defaulting to user installation because normal site-packages is not writeable
ERROR: Ignored the following versions that require a different python version: 2.82.1 Requires-Python >=3.7, <3.8
ERROR: Could not find a version that satisfies the requirement bpy==4.1.0 (from versions: 3.4.0, 3.5.0, 3.6.0, 4.0.0)
ERROR: No matching distribution found for bpy==4.1.0
```

Any idea? I will try the symlink or vsc extension for development. thanks for the tip

**Reply by merlinND** [reply]
Hello @tadz-io,

I have very limited experience with Blender extensions, but based on `Calling operator "bpy.ops.import_scene.mitsuba" error, could not be found`: are you sure that the extension is really installed and enabled?
You can check by creating and exporting a simple scene (e.g. just the default cube).

As @bathal1 said, trying to install `bpy` with `pip` will not help, you should use the `bpy` version that comes bundled with your Blender version.

**Reply by tadz-io** [reply]
@merlinND sorry, i should have been more explicit. I got it to work now by using VSC's blender extension as suggested by @bathal1 , and running it from my local mitsuba-blender repo. I actually had to the delete the installed mitsuba-blender extension in blender for it work (or rename it). Thanks for the input though
